### PR TITLE
feat: Analytics to aggregate data exchange service [DHIS2-13105]

### PR DESCRIPTION
**Pull request by larshelge**
### Overview

Introduces a new data exchange service. The source is the analytics API data value set format, and the destination is the data value set API. The destination can be "internal", meaning the this instance of DHIS 2, or "external", meaning a remote instance of DHIS 2.

Introduces a new entity `AnalyticsDataExchange` and a new maven project `dhis-service-data-exchange`.

The benefits of the data exchange service compared to a custom script solution are:

* Can be defined as metadata, and can be imported-exported in metadata and metadata packages.
* Better performance, run internally using Java APIs, not over the network and the web API.
* Can be scheduled as a DHIS 2 job.
* Runs in parallel for better performance.
* Auto-partitioning of program indicators for less memory usage.
* Encryption of passwords/API tokens, more secure.

### ToDo

* Encryption of DHIS 2 password
* Scheduling of data exchange jobs
* Parallel execution
* Auto-partitioning of data items
* PAT authentication support

### API

Create exchange:

```
POST /api/analyticsDataExchanges
```

Get, update and delete exchange:

```
GET PUT DELETE /api/analyticsDataExchanges/{id}
```

Run exchange by ID:

```
POST /api/analyticsDataExchange/{id}/exchange
```

Example of internal exchange:

```json
{
    "name": "Internal data exchange",
    "source": {
        "requests": [
            {
                "dx": [
                    "fbfJHSPpUQD",
                    "cYeuwXTCPkU",
                    "Jtf34kNZhzP"
                ],
                "pe": [
                    "LAST_12_MONTHS",
                    "202201"
                ],
                "ou": [
                    "ImspTQPwCqd"
                ],
                "inputIdScheme": null,
                "outputIdscheme": null
            }
        ]
    },
    "target": {
        "type": "INTERNAL",
        "request": {
            "dataElementIdScheme": null,
            "orgUnitIdScheme": null,
            "categoryOptionComboIdScheme": null,
            "idScheme": null
        }
    }
}
```

Example of external exchange:

```json
{
    "name": "External data exchange",
    "source": {
        "requests": [
            {
                "dx": [
                    "fbfJHSPpUQD",
                    "cYeuwXTCPkU",
                    "Jtf34kNZhzP"
                ],
                "pe": [
                    "LAST_12_MONTHS",
                    "202201"
                ],
                "ou": [
                    "ImspTQPwCqd"
                ],
                "inputIdScheme": null,
                "outputIdscheme": null
            }
        ]
    },
    "target": {
        "type": "EXTERNAL",
        "api": {
            "url": "https://play.dhis2.org/2.38.0",
            "username": "admin",
            "password": "district"
        },
        "request": {
            "dataElementIdScheme": null,
            "orgUnitIdScheme": null,
            "categoryOptionComboIdScheme": null,
            "idScheme": null
        }
    }
}
```

### CODE REVIEW

**Review by mortenoh — APPROVED** [review]

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataexchange/analytics/AnalyticsDataExchange.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+public class AnalyticsDataExchange
+    extends BaseIdentifiableObject
+    implements MetadataObject
+{
+    private Source source;
+
+    private Target target;
+
+    @JsonProperty
+    @JacksonXmlProperty( namespace = DxfNamespaces.DXF_2_0 )
```
[the remark this answers is not in the document]
Comment (larshelge) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataexchange/analytics/AnalyticsDataExchange.java:48]: Great point.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataexchange/analytics/AnalyticsDataExchange.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    public Source getSource()
+    {
+        return source;
+    }
+
+    public void setSource( Source source )
```
[the remark this answers is not in the document]
Comment (larshelge) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataexchange/analytics/AnalyticsDataExchange.java:54]: Yep.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataValueSet.java [hunk 3]
```diff
@@ -50,7 +50,7 @@
 @Setter
 @NoArgsConstructor
 @JacksonXmlRootElement( localName = "dataValueSet", namespace = DxfNamespaces.DXF_2_0 )
-public final class DataValueSet
+public class DataValueSet
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataValueSet.java:53]: Is there a particular reason to remove the `final`? I did not see a new subclass and having worked hard to get rid of them I would like to keep it that way :D

**Review by jbee — COMMENTED** [review]
I wonder if it makes sense to have all classes of the request details (`Source`, `Target`, `Filter`...) modelled as inner classes of the outermost record. They are used together and if one looks into it it helps to look at all of them together while having them split into many small simple files has no benefit but hinders readability. Same for the response.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/feedback/ErrorCode.java [hunk 4]
```diff
@@ -276,6 +276,10 @@
     E6210( "Restore expiration date is not set for user account" ),
     E6211( "User account restore invitation has expired" ),
 
+    /* Data exchange */
+    E6300( "DHIS 2 client request failed: {0} {1}" ),
+    E6301( "Analytics data change not found or not accessible: `{0}`" ),
```
[the remark this answers is not in the document]
Comment (larshelge) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/feedback/ErrorCode.java:281]: Supposed to be "exchange".

File: dhis-2/dhis-services/dhis-service-data-exchange/src/main/java/org/hisp/dhis/dataexchange/client/Dhis2Client.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+import org.springframework.web.util.UriComponentsBuilder;
+
+/**
+ * Client for DHIS 2.
+ *
+ * @author Lars Helge Overland
+ */
+@Slf4j
+@Getter
+public class Dhis2Client
```
[the remark this answers is not in the document]
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-data-exchange/src/main/java/org/hisp/dhis/dataexchange/client/Dhis2Client.java:65]: Right, this is intended as a generic client. We can refactor it later if we see the need for more specialized clients, or if someone else wants to take the Dhis2Client name.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataValueSet.java (hunk 3 above)
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataValueSet.java:53]: Right sorry for that, it was for mocking and testing purposes. Mockito does not allow mocking of final classes. There could be a better way to do this, any ideas?

Comment (jbee) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataValueSet.java:53]: No, that is fine. Just wanted to know if there are actual sub-classes.

**Review by jbee — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataValueSet.java (hunk 3 above)
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataValueSet.java:53]: Understand.

**Pull request merged**